Fixed #26033 -- Added argon2 password hasher.

bwesterb opened the pull request:
argon2 is the winner of the [Password Hashing Competition](https://password-hashing.net).

1. There are two python bindings available for argon2:  `argon2` and `argon2-cffi`.  The first uses `ctypes` and the latter `cffi`.  As `cffi` works for most environments, I choose that one.
2. `argon2` has four parameters: `time_cost`, `memory_cost`, `parallelism` and `type`.  I choose to copy the defaults of `argon2-cffi`.  [Section 9 of the argon2 specification discusses recommended  parameters.](https://password-hashing.net/argon2-specs.pdf)

https://code.djangoproject.com/ticket/26033#ticket

timgraham commented:
Thanks for the patch. Design decisions (such as the choice of what algorithms to support in core) are usually made on the [django-developers mailing list](https://groups.google.com/forum/#!forum/django-developers). I would post this idea there to get feedback.

bwesterb commented:
Ok thanks. [Post.](https://groups.google.com/forum/#!topic/django-developers/NTfqP4eNVyA)

hynek commented:
Since Florian mentioned he’d like some feedback from me or @dstufft on the mailing list:

Sadly I can’t really comment on the Django portions *but* the usage of Argon2 libraries is rather straight forward as long as you stay away from the `argon2.low_level.core()` binding.  The authors mention in the paper that no parameter value is actually insecure.

I have two minor general comments:

- I wouldn’t test with a time cost of 8. :)  If you want to check whether arguments get propagated, I would recommend to go down to 1 and not up lest you wanna slow down your test suite.
- If you want to keep using `argon2-cffi`’s defaults, you can also just use the relevant [constants](http://argon2-cffi.readthedocs.org/en/stable/api.html#argon2.DEFAULT_TIME_COST).

bwesterb commented:
* I decreased the time cost for the test from 8 to 1 as suggested by @hynek.
* I am ok with using `argon2` constants directly.  However, this would be different from the other `PasswordHasher`s that hardcode their default.  We can always change this after it has been merged.

This PR is ready to be merged.

apollo13 commented on setup.py:
```diff
@@ -49,6 +49,7 @@
     ]},
     extras_require={
         "bcrypt": ["bcrypt"],
+        "argon2": ["argon2-cffi >= 16.0.0"],
```
Curious, what did change in version 16, so you have to require it?

Rotund replied:
16 added the argon2.low_level module which is used to set a salt. (See comments below in the review.)

bwesterb replied:
Indeed.

apollo13 commented:
Is there a specific reason to go with the low_level api?

apollo13 commented on django/contrib/auth/hashers.py:
```diff
@@ -515,3 +515,67 @@ def safe_summary(self, encoded):
+                                            time_cost=self.time_cost,
+                                            memory_cost=self.memory_cost,
+                                            parallelism=self.parallelism,
+                                            hash_len=16,
```
Please use ` argon2.DEFAULT_HASH_LENGTH` here,

bwesterb replied:
Done.  We do have to be careful: if `argon2.DEFAULT_HASH_LENGTH` is increased significantly in the future, `encode` can return a string longer than the maximum 128.

aaugustin commented on django/conf/global_settings.py:
```diff
@@ -492,6 +492,7 @@
 PASSWORD_HASHERS = [
     'django.contrib.auth.hashers.PBKDF2PasswordHasher',
     'django.contrib.auth.hashers.PBKDF2SHA1PasswordHasher',
+    'django.contrib.auth.hashers.Argon2PasswordHasher',
```
What is the reason for adding this hasher it after PBKDF2 but before BCrypt?

Adding it anywhere but in first position will have no effect anyway. It would only serve to upgrade existing Argon-hashed passwords to PBKDF2e

bwesterb replied:
If it is not in the list, the tests won't run.

The position in the list is meant to suggest it's the second-best option after PBKDF2.

timgraham replied:
Sorry for ignorance, but if PBKDF2 is better, what's the value in adding something "worse"? p.s. the patch is missing documentation -- presumably you could explain it there.

dstufft replied:
From a security stand point, argon2 is better than PBKDF2 because it's memory hard as well as CPU hard. Security wise argon2 > scrypt > PBKDF2 ~= bcrypt.

dstufft replied:
Of course, the above doesn't take into account the availability of the hasher, since PBKDF2 is available (via efficient C) in the stdlib in 2.7.9+ and 3.something+ and can be done reasonably in pure Python before that whereas scrypt, bcrypt, and argon2 will all require an external library (generally C based).

timgraham replied:
So it should be first? Otherwise, I agree with Aymeric's point that it wouldn't ever be used. If it's required to make the tests run but shouldn't be in the defaults, then please use `@override_settings()` in the tests.

apollo13 replied:
Our code currently requires the first hasher to be usable, putting it as first and not installing the extension will break (look at `get_hasher('default')`)

dstufft replied:
Surely you want all of the password hashes in the list of available ones so that people can log in with them? Of course argon2 is unlikely to be out there in the wild, but if someone tried it and then decided to switch back they wouldn't be able to do so by deleting the settings, they'd have to copy/paste the default and re-add it. There's little harm in including it by default afaik.

You might not want to have something other than PBKDF2 to be the default though, people might accidently have bcrypt or argon2 installed and not realize it and end up unable to log in if they deploy. Not sure if that's a big worry or not though.

carljm replied:
I think this ordering is reasonable. PBKDF2 should be the default preferred hasher, because it is always usable without requiring a third-party package. All of the bundled hashers should be in the default list somewhere, even though it has no real immediate effect, both because it provides additional "documentation" of their existence, and for the reason @dstufft noted (if someone tries out the argon2 hasher in a custom hashers setting and then deletes their custom setting, those passwords should still be upgradable via the default setting).

apollo13 replied:
I think the best argument is "There's little harm in including it by default afaik.". If someone tried using argon2 they can as well adjust their settings, after all they are testing against an unreleased version.

dstufft replied:
Presumably at some point this change will be released.

bwesterb replied:
1.  On compatibility.   The hasher is not as compatible as PBKDF2, but it is close.  `argon2-cffi`, the library used in the PR, comes with `libargon2` bundled with which it binds with cffi instead of ctypes.  This means it will work just fine on lots of platforms including pypy.  It will exclude, for instance, Google App Engine.
2. On security.  The panel of the Password Hashing Competition recommends the use of argon2 over PBKDF2, which it considers a legacy algorithm.   The only downside I see, is its youth.

I think it would be great to have argon2 in Django as an option, such that those who are wary of PBKDF2 can switch to argon2 with minimal effort.

Maybe we could switch to argon2 as default in a later release.  Some questions:

1. For ease of deployment, `argon2-cffi` could be bundled with Django?
2. Wouldn't it be possible to make `get_hasher` find the first available hasher, to support older platforms.
3. I could write a pure python version of `argon2-cffi` that only implements verification, such that platforms that become incompatible can move away to another hasher?

carljm replied:
Making the hasher-selection smart enough to discard unusable hashers from the top of the list seems like a reasonable improvement, but out of scope for this PR. For the purposes of this PR, I think adding argon2 after PBKDF2 in the default list (as you have it now) is perfectly reasonable.

apollo13 replied:
On Tue, Jan 5, 2016 at 8:52 PM Carl Meyer <notifications@github.com> wrote:

> Making the hasher-selection smart enough to discard unusable hashers from
> the top of the list seems like a reasonable improvement


Security wise I do not like that idea. For example: Argon2 is the first in
the list for one of your deployments, you setup a new server (and switch to
it) which accidentally misses argon2 libs and your passwords get
downgraded… Chances that you will notice that anytime soon are very slim
imo.

carljm replied:
Yeah on second thought I agree. The security benefit of moving a better hasher that requires a third-party install to the top of the default list is minimal or nonexistent, since you still have to explicitly install the third-party requirement to gain the benefit, at which point you may as well have to set the setting as well. Automatic changes in behavior based only on the presence or absence of a third-party package are questionable in general.

bwesterb replied:
Wouldn't bundling argon2-cffi (and thus also libargon2) solve part of it?  Then you only have to make sure the build is forced, when available.

dstufft replied:
I think (hope?) Python packaging is at a point that if Django wanted to make argon2 the default that it could just depend on argon2-cffi instead of bundling. That being said though, the first mandatory dependency being a C based one might be ambitious, and there isn't much reason to worry about having PBKDF2 as a default. PBKDF2 is considered secure and is a perfectly fine password KDF.

bwesterb replied:
If we would all pick sufficiently strong password, than `SHA256` would be a perfectly fine password hash.  The current default setting for PBKDF2 on Django, uses 60,000 iterations of `SHA256`.  Thus if an attacker would have been able to crack your stolen SHA256 hashes in a day, he would have to wait 164 years with PBKDF2.

A capable attacker will use a GPU, a FPGA or maybe even an ASIC to brute force PBKDF2.  Per dollar an ASIC is in the order of 50,000 times more efficient for SHA256, than a CPU.  (Check MHash/J between [ASIC bitcoin miners](https://en.bitcoin.it/wiki/Mining_hardware_comparison) and [CPU bitcoin miners](https://en.bitcoin.it/wiki/Non-specialized_hardware_comparison).)  argon2 (and the other new password hashes) try to close this gap: they are designed such that they shouldn't be a lot faster to compute on an ASIC than on a general purpose CPU.  Thus, if there are no weaknesses in argon2, and we take parameters such that it runs as quick as PBKDF2, then a password which could have been brute forced in a day if PBKDF2 was used, would take in the order of centuries with argon2.

I do agree: PBKDF2 is a fine password hash.  However, I wanted to emphasise that argon2 promises to be a real improvement instead of a marginal one.

bwesterb commented:
@apollo13 `argon2-cffi`'s high-level API (`argon2.PasswordHasher`) does not allow one to specify a salt (on purpose).  It generates a new random salt itself.  As Django's API wants to generate a salt itself, we use the lower-level API.

timgraham commented on django/contrib/auth/hashers.py:
```diff
@@ -515,3 +515,67 @@ def safe_summary(self, encoded):
+
+    def encode(self, password, salt):
+        argon2 = self._load_library()
+        data = argon2.low_level.hash_secret(force_bytes(password),
```
Please use 4 space hanging indent: 
```
argon2.low_level.hash_secret(
    force_bytes(password),
    ...,
)
```

bwesterb replied:
Done.

timgraham commented on django/contrib/auth/hashers.py:
```diff
@@ -515,3 +515,67 @@ def safe_summary(self, encoded):
+        algorithm, data = encoded.split('$', 1)
+        assert algorithm == self.algorithm
+        try:
+            ok = argon2.low_level.verify_secret(force_bytes('$' + data),
```
no need for "ok" variable.. can return directly.

timgraham commented on django/contrib/auth/hashers.py:
```diff
@@ -515,3 +515,67 @@ def safe_summary(self, encoded):
+
+    def safe_summary(self, encoded):
+        algorithm, variety, raw_pars, salt, data = encoded.split('$', 5)
+        pars = dict([bit.split('=', 1) for bit in raw_pars.split(',')])
```
don't think you need the explicit list comprehension (drop braces)

bwesterb replied:
Thanks.

timgraham commented on django/contrib/auth/hashers.py:
```diff
@@ -515,3 +515,67 @@ def safe_summary(self, encoded):
+        pars = dict([bit.split('=', 1) for bit in raw_pars.split(',')])
+        assert algorithm == self.algorithm
+        assert len(pars) == 3 and 't' in pars and 'm' in pars and 'p' in pars
+        return (self.time_cost != int(pars['t']) or
```
as above:
```
return (
    ... or 
    ...
)
```

bwesterb replied:
Done

timgraham commented on tests/auth_tests/test_hashers.py:
```diff
@@ -209,6 +214,53 @@ def setter(password):
         finally:
             hasher.rounds = old_rounds
 
+    @skipUnless(argon2, "argon2-cffi not installed")
```
Could you put these in their own test class so we only need skipUnless on the class?

bwesterb replied:
Should I also do it for bcrypt?  (I copied it from there.)

timgraham replied:
Not necessary for this. It could be done as a separate PR if you are so inclined.

bwesterb replied:
Done.  (I left bcrypt as it is.)

timgraham commented on tests/requirements/base.txt:
```diff
@@ -1,3 +1,4 @@
+argon2-cffi >= 16.0.0
```
these dependencies are also mentioned in docs/internals/contributing/writing-code/unit-tests.txt

bwesterb replied:
Done

bwesterb commented:
@timgraham I saw you set the "missing documentation" label on the issue tracker.  Could you help me out and tell me which documentation I should write?

timgraham commented:
docs/topics/auth/passwords.txt -- see the similar docs for bcrypt there.

timgraham commented on django/contrib/auth/hashers.py:
```diff
@@ -515,3 +515,70 @@ def safe_summary(self, encoded):
+        algorithm, data = encoded.split('$', 1)
+        assert algorithm == self.algorithm
+        try:
+            return argon2.low_level.verify_secret(force_bytes('$' + data),
```
As above, please use hanging indent.

bwesterb replied:
Done.

bwesterb commented:
@timgraham Thanks.  I added the documentation and release notes.    Please feel free to change the wording, tone and contents as you like.

Before you merge, I will squash the commits.  (For now it is easier to keep track of what changed in the PR.)

By the way, `UnsaltedSHA1PasswordHasher` and `UnsaltedMD5PasswordHasher` were missing in the documentation.  Was this on purpose.  I've added them for the moment, though.

bwesterb commented:
Ok, squashed it already.

timgraham commented:
buildbot, add to whitelist.

aaugustin commented on docs/releases/1.10.txt:
```diff
@@ -55,6 +55,31 @@ Minor features
+          'django.contrib.auth.hashers.CryptPasswordHasher',
+      ]
+
+  You need to keep the other entries in this list, or else Django won't
```
"the other entries" isn't entirely accurate; very few websites well need to keep anything other than `PBKDF2PasswordHasher`, the default hasher since Django got its current password hashing mechanism.

I would say:

      PASSWORD_HASHERS = [
          'django.contrib.auth.hashers.Argon2PasswordHasher',
          'django.contrib.auth.hashers.PBKDF2PasswordHasher',
    ]

If your database contains password encoded with other hashers, you need to add them to the list, or else the corresponding users won't be able to log in. See the documentation of :setting:`PASSWORD_HASHERS` for details.

bwesterb replied:
Some people might blindly copy and paste the displayed code to enable Argon2, without reading the remark after it.  It does not hurt to list all hashers, so that seems to me to be the safe choice.

This is also how the documentation is written to enable BCrypt.  See `topics/auth/passwords.txt`.

Off course, I could change it everywhere.

aaugustin replied:
Keeping it consistent is a reasonable choice for now. Leave it as is.

timgraham replied:
I would link to "argon2_usage" instead of repeating the example usage parts of it in the release notes.

bwesterb replied:
> I would link to "argon2_usage" instead of repeating the example usage parts of it in the release notes.

Done.

aaugustin commented on docs/topics/auth/passwords.txt:
```diff
@@ -60,17 +60,63 @@ The default for :setting:`PASSWORD_HASHERS` is::
+
+        PASSWORD_HASHERS = [
+            'django.contrib.auth.hashers.Argon2PasswordHasher',
+            'django.contrib.auth.hashers.PBKDF2PasswordHasher',
```
As above.

hynek commented on docs/releases/1.10.txt:
```diff
@@ -55,6 +55,31 @@ Minor features
+* Support for the Argon2_ password hash has been added.  Argon2 is the winner
+  of the recent `Password Hashing Competition`_, which has been organized
+  by the community in similar style as the NIST competitions that selected
+  the hash SHA-3 and cipher AES.
```
I find this a bit misleading.  It kind of associates SHA-3/AES with Argon2 although only the method (= competition) stayed same.

It would word it more like “…in similar style as the NIST competitions that chose cryptographic standards in the past.”

bwesterb replied:
I don't completely understand your objection.  You are afraid we oversell argon2?  Would it help if we add a note that it is a young algorithm?

BTW, a lot of people probably wouldn't know about the NIST competitions.  So if you want to leave out AES and SHA-3, then maybe it's better to completely leave out the reference to NIST.

hynek replied:
I’m not afraid to oversell it, I’m just not partial to name dropping of unrelated technologies.  But it’s up to the Django folks to make that call; it sure is illustrative.

hynek commented on docs/topics/auth/passwords.txt:
```diff
@@ -60,17 +60,63 @@ The default for :setting:`PASSWORD_HASHERS` is::
+
+Argon2_ is the winner of the Password Hashing Competition (PHC_), a community
+organized open competition similar to those organized by NIST_ to select the
+hash SHA-3 and the symmetric cipher AES.  Argon2_, similar to other new
```
same

bwesterb commented:
All comments have been addressed.

bwesterb commented:
@timgraham Everything OK?

timgraham commented on docs/topics/auth/passwords.txt:
```diff
@@ -60,17 +60,63 @@ The default for :setting:`PASSWORD_HASHERS` is::
         'django.contrib.auth.hashers.BCryptPasswordHasher',
         'django.contrib.auth.hashers.SHA1PasswordHasher',
         'django.contrib.auth.hashers.MD5PasswordHasher',
+        'django.contrib.auth.hashers.UnsaltedSHA1PasswordHasher',
```
Thanks for making this correction, however, it should be done in a separate commit. I was looking at doing that but wondered if we need to keep these hashers around: https://groups.google.com/d/topic/django-developers/ZeRJU8YVrxg/discussion

bwesterb replied:
I reverted the corrections in this PR.

timgraham commented on django/contrib/auth/hashers.py:
```diff
@@ -515,3 +515,72 @@ def safe_summary(self, encoded):
         ])
+
+
+class Argon2PasswordHasher(BasePasswordHasher):
```
The hashers seem to be ordered based on where they appear in settings so I'd move it up.

bwesterb replied:
Done.

timgraham commented on tests/auth_tests/test_hashers.py:
```diff
@@ -359,3 +364,53 @@ def test_load_library_importerror(self):
+        self.assertTrue(check_password('', blank_encoded))
+        self.assertFalse(check_password(' ', blank_encoded))
+
+    def test_argon2_upgrade(self):
```
Add tests for changing the other parameters that `must_update` affects?

bwesterb replied:
Done.

timgraham commented on docs/topics/auth/passwords.txt:
```diff
@@ -138,14 +183,19 @@ Increasing the work factor
-you'll subclass the appropriate algorithm and override the ``iterations``
-parameters. For example, to increase the number of iterations used by the
-default PBKDF2 algorithm:
+iterations needs to be increased.  Argon2 has two additional knobs that can
```
I'm going to suggest some edits to move argon2 discussion to a separate section with a few todos that I hope can complete: https://dpaste.de/aaF5  You can apply the patch with `patch -p1 -i raw.diff`. Thanks, this is coming along!

bwesterb replied:
Done.  I added some text on choosing parameters.

timgraham commented on docs/topics/auth/passwords.txt:
```diff
@@ -173,10 +216,32 @@ default PBKDF2 algorithm:
+#. ``parallelism`` controls how many CPUs the computation of the hash can be
+   parallelized on.
+
+The default values of these attributes are probably fine for you.  If you
```
Should we say what the defaults are? Should we consider increasing the defaults for each major release of Django like we do for PBKDF2? https://docs.djangoproject.com/en/dev/internals/howto-release-django/#new-stable-branch-tasks
Please use 1 space between sentences.

bwesterb replied:
It is a good idea to review the defaults for each major release.  The analogue of the current rule would be to increase `memory_cost` with approximately 20% for each major release.  Maybe it would be a better idea to use actual hardware as a reference: adjust the parameters such that the hash takes a certain amount of milliseconds on the cheapest Amazon EC2 instance.

bwesterb replied:
Maybe coordinate changing the defaults with @hynek ?

hynek replied:
The defaults aren’t too likely to change very frequently since even the lowest settings are “safe enough”.  Choosing the target verification time is highly individual so I guess the best approach for Django is to decide on a time and tune all checkers to meet it so people don't have to think about it?  My own target was roughly 0.5ms–1ms “on my notebook”.

timgraham commented on docs/topics/auth/passwords.txt:
```diff
@@ -173,10 +216,32 @@ default PBKDF2 algorithm:
+#. Adjust ``time_cost`` and measure the time hashing a password takes.
+   Pick a ``time_cost`` that takes an acceptable time for you.
+   If ``time_cost`` set to 1 is unacceptably slow, lower ``memory_cost``.
+
```
only 1 blank line needed

bwesterb replied:
Done.

timgraham commented on docs/topics/auth/passwords.txt:
```diff
@@ -68,9 +69,46 @@ The default for :setting:`PASSWORD_HASHERS` is::
+------------------------
+
+Argon2_ is the winner of the 2015 `Password Hashing Competition`_, a community
+organized open competition to select a next generation hasing algorithm. It's
```
sorry my typo: hashing

bwesterb replied:
Done.

timgraham commented on docs/topics/auth/passwords.txt:
```diff
@@ -173,10 +216,32 @@ default PBKDF2 algorithm:
+
+Argon2 has three attributes that can be customized:
+
+#. ``time_cost`` controls the number of iterations within the hash
```
add period after each sentence

bwesterb replied:
Done.

timgraham commented on tests/auth_tests/test_hashers.py:
```diff
@@ -359,3 +364,60 @@ def test_load_library_importerror(self):
+        self.assertEqual('argon2', hasher.algorithm)
+
+        for attr, summary_key, new_value in (
+                ('time_cost', 'time cost', 1),
```
I'd declare this as a `test_params` variable above to avoid the funky indent.

timgraham replied:
Another option could be to refactor into 3 separate test methods that call a common helper method to run the logic currently in the loop. This can be easier to debug than assertions that run within a loop.

bwesterb replied:
I split it into two functions: one test method and one helper.  (Without the loop, you can see the invocation in the backtrace.)  Is that ok?

timgraham commented on tests/auth_tests/test_hashers.py:
```diff
@@ -359,3 +364,60 @@ def test_load_library_importerror(self):
+                    state['upgraded'] = True
+
+                # Check that no upgrade is triggered.
+                self.assertTrue(check_password('letmein', encoded, setter,
```
1 line is okay --- we prefer longer lines up to 119 characters when it helps readability.

bwesterb replied:
Done.

timgraham commented:
I'd like to merge https://github.com/django/django/pull/6103 first, then rebase the docs for the changes there.

bwesterb commented:
Just let me know when you merged it.  Then I will rebase.

bwesterb commented:
@timgraham  Rebased.

timgraham commented:
Could you add a `harden_runtime()` method based on the patch in today's security release? 67b46ba7016da2d259c1ecc7d666d11f5e1cfaab

bwesterb commented:
**The problem.** The runtime of PBKDF2 is practically linear in the number of iterations.  For argon2, it is more complex.  If I am not mistaken, it's runtime is proportional to

    time_cost * memory_cost / max(available_threads, parallelism) + memory_cost * constant1 + constant2

The first term is for the main iteration of argon2, which goes over all blocks of memory `time_cost` times.  The second is for the compression of all memory into one final block (`B_final` in the spec.) at the end whose hash will be the result of argon2.  The third is for the computation of the initial hash of the parameters (`H_0` in the spec.)

Suppose we double the `time_cost`. Then argon2 will run just shy of twice as fast: the second and the third term are the difference.  Here is some simple code that verifies this: https://gist.github.com/bwesterb/fe0e8be93c978aaa911d  The difference is even bigger than I expected: 10%.

**What to do?**
I see a few possibilities

1. Use a no-op `harden_runtime` and live with the issue.
3. Let Django compute the expected runtime of the default password hash beforehand and let the harden_runtime sleep (or spinwait) the difference.
3. Instead of changing the parameters of argon2, simply iterate argon2 itself.  (In other words: use `PBKDF2(argon2)`)

The last would be the best mitigation of the issue at hand, but would over time give us a hash which is a lot weaker than it could have been.  Option 2 would also work if a hash is replaced by a different, but slower hash (instead of just when the parameters are changed).

apollo13 commented:
I am leaning towards option one. The security fix is mainly for pbkdf2
where we constantly increase iterations. I do not think that algos like
argon 2 need such an aggressive approach (I have to read up on argon
though, but for bcrypt adding 1 to the work factor means doubling the
workload -- not something you'd do lightly)

On Wed, Mar 2, 2016, 14:54 Bas Westerbaan <notifications@github.com> wrote:

> *The problem.* The runtime of PBKDF2 is practically linear in the number
> of iterations. For argon2, it is more complex. If I am not mistaken, it's
> runtime is proportional to
>
> time_cost * memory_cost / max(available_threads, parallelism) + memory_cost * constant1 + constant2
>
> The first term is for the main iteration of argon2, which goes over all
> blocks of memory time_cost times. The second is for the compression of
> all memory into one final block (B_final in the spec.) at the end whose
> hash will be the result of argon2. The third is for the computation of the
> initial hash of the parameters (H_0 in the spec.)
>
> Suppose we double the time_cost. Then argon2 will run just shy of twice
> as fast: the second and the third term are the difference. Here is some
> simple code that verifies this:
> https://gist.github.com/bwesterb/fe0e8be93c978aaa911d The difference is
> even bigger than I expected: 10%.
>
> *What to do?*
> I see a few possibilities
>
>    1. Use a no-op harden_runtime and live with the issue.
>    2. Let Django compute the expected runtime of the default password
>    hash beforehand and let the harden_runtime sleep (or spinwait) the
>    difference.
>    3. Instead of changing the parameters of argon2, simply iterate argon2
>    itself. (In other words: use PBKDF2(argon2))
>
> The last would be the best mitigation of the issue at hand, but would over
> time give us a hash which is a lot weaker than it could have been. Option 2
> would also work if a hash is replaced by a different, but slower hash
> (instead of just when the parameters are changed).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/django/django/pull/5876#issuecomment-191246555>.
>

bwesterb commented:
@timgraham I added a no-op `harden_runtime` as suggested by @apollo13

timgraham commented:
merged in b4250ea04a88f6c4fdf84dc8624baa1cf3e0f568, thanks!

timgraham commented:
We have a test failure with the release of argon2_cffi 16.1 which [changes the hash](https://github.com/hynek/argon2_cffi/commit/9380662da1e54adbbccc403cbb4fd0300c0ee2ae) and adds an extra dollar sign which causes problems with Django trying to split it. How do you think we should proceed?

hynek commented:
Argon2 has a new version 1.3 that introduced a version tag. 

My suggestions:

- fix the hash in your tests.
- you can check what version the hash is and use that information for opportunistic re-hashing

JFTR Hashes are backward compatible.   You can verif old hashes with the new release.

bwesterb commented:
I will write a patch for this this week.

hynek commented:
(Oh and I didn't change anything; it's the standard/official bindings that evolved.)

bwesterb commented:
I created a PR: #6489
